# Sentiment with Flair

Flair offers models that we can use out-of-the-box. One of those is the English sentiment model, which we will learn how to use here.

First, we need to make sure Flair has been installed, we do this in our CLI with:

```
pip install flair
```

Flair uses PyTorch/TensorFlow in under the hood, so it's essential that you also have one of the two libraries (or both) installed. There are a few steps in applying sentiment analysis, these are:

1. Initializing the model.

2. Tokenizing input text.

3. Processing with the model.

4. *(Optional) Formatting the outputs.*

We then load the English sentiment model like so:

In [4]:
# ! pip install flair
import flair
model = flair.models.TextClassifier.load('en-sentiment')

The first time this `load` method is run for the `'en-sentiment'` model the model will be downloaded. After this, the model is initialized. The `en-sentiment` model is a distilBERT model fitted with a classification head that outputs two classes - negative and positive.

Our next step is to tokenize input text. For this we use the Flair `Sentence` object, which we initialize by passing our text into it:

In [5]:
text = "I like you. I love you"  # we are expecting a confidently positive sentiment here

sentence = flair.data.Sentence(text)

sentence

Sentence[7]: "I like you. I love you"

Here we now have the Flair `Sentence` object, which contains our text, alongside a *tokenized* version of it (each word/punctuation character is an individual token):

In [6]:
sentence.to_tokenized_string()

'I like you . I love you'

The next step is to process our tokenized inputs through out distilBERT classifier:

In [10]:
model.predict(sentence)

The `predict` method doesn't output our prediction, instead the predictions are added to our `sentence`:

In [11]:
sentence

Sentence[7]: "I like you. I love you" → POSITIVE (0.9933)

Here we can see that we are predicting a `POSITIVE` sentiment with a probability of `0.9933`, which is **very** confident as expected. Let's repeat the process with something more negative.

In [12]:
text = "I hate it when I'm not learning"
sentence = flair.data.Sentence(text)
model.predict(sentence)
sentence

Sentence[8]: "I hate it when I'm not learning" → NEGATIVE (0.9991)

In [19]:
# help(model.predict(sentence))

And we correctly predict a `NEGATIVE` sentiment. Finally, we will typically want to extract our predictions and format them into the format that we need for our own use-case (for example plotting sentiment over time). Let's take a look at how we do that.

The `Sentence` object provides us with a method called `get_labels`, we can use this to extract our sentiment prediction.

In [13]:
sentence.get_labels()

['Sentence[8]: "I hate it when I'm not learning"'/'NEGATIVE' (0.9991)]

From this method we actually get a list, which contains our label object. To access each item in the list we need to dig a little deeper. We first access the label object by accessing the *0th* index of our list. Flair `Label` objects contain two attributes, `score` and `value` - which contain our prediction.

In [14]:
sentence.get_labels()[0]

'Sentence[8]: "I hate it when I'm not learning"'/'NEGATIVE' (0.9991)

In [20]:
help(sentence.get_labels()[0])

Help on Label in module flair.data object:

class Label(builtins.object)
 |  Label(data_point: 'DataPoint', value: str, score: float = 1.0)
 |  
 |  This class represents a label. Each label has a value and optionally a confidence score. The
 |  score needs to be between 0.0 and 1.0. Default value for the score is 1.0.
 |  
 |  Methods defined here:
 |  
 |  __eq__(self, other)
 |      Return self==value.
 |  
 |  __hash__(self)
 |      Return hash(self).
 |  
 |  __init__(self, data_point: 'DataPoint', value: str, score: float = 1.0)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __lt__(self, other)
 |      Return self<value.
 |  
 |  __repr__(self)
 |      Return repr(self).
 |  
 |  __str__(self)
 |      Return str(self).
 |  
 |  set_value(self, value: str, score: float = 1.0)
 |  
 |  to_dict(self)
 |  
 |  ----------------------------------------------------------------------
 |  Readonly properties defined here:
 |  
 |  labeled_identifier
 |  
 

In [21]:
sentence.get_labels()[0].score

0.9991464614868164

In [17]:
sentence.get_labels()[0].value

'NEGATIVE'

Alternatively, we can access the label values directly (although not recommended) like so:

In [22]:
sentence.labels[0].score, sentence.labels[0].value

(0.9991464614868164, 'NEGATIVE')